packages/components/badge/src/Badge/Badge.tsx

```diff
@@ -5,49 +5,83 @@ import {
   type CommonProps,
   type PropsWithHTMLElement,
   type ExpandProps,
+  type PickUnion,
 } from '@contentful/f36-core';
 
 import type { BadgeSize, BadgeVariant } from '../types';
 import { getBadgeStyles } from './Badge.styles';
 
-export interface BadgeInternalProps extends CommonProps {
-  /**
-   * Sets the size of the component
-   * @default default
-   */
-  size?: BadgeSize;
-  /**
-   * Determines the variation of the component
-   * @default primary
-   */
-  variant?: BadgeVariant;
-
-  children: React.ReactNode;
-}
+type BadgeSizeWithIconProps =
+  | {
+      /**
+       * Sets the size of the component
+       * @default default
+       */
+      size?: PickUnion<BadgeSize, Exclude<BadgeSize, 'small'>>;
+      /**
+       * Expects any of the icon components. Renders the icon aligned to the start
+       */
+      startIcon?: React.ReactElement;
+      /**
+       * Expects any of the icon components. Renders the icon aligned to the end
+       */
+      endIcon?: React.ReactElement;
+    }
+  | {
+      /**
+       * Sets the size of the component
+       * @default default
+       */
+      size: PickUnion<BadgeSize, 'small'>;
```

Comment:
```suggestion
      size?: BadgeSize;
```

Reply:
We can't do that sadly, because we want to avoid the usage of the icons when the size is set to `small`, which is why we are using this discriminated union type, so in case the user sets `size="small"` and tries to add an icon TS will show an error

Reply:
A comment here might be helpful for future human browsers